File: src/morie/fn/sgtspc.py
# morie.fn -- function file (rootcoder007/morie)
"""Spectrum of the normalised Laplacian."""

from . import _tail1core as C

from ._richresult import RichResult

__all__ = ["lapspec", "sgt_spectrum"]


def lapspec(W, tol=1e-10):
    """Eigenvalues of the normalised Laplacian, in increasing order.

    Two facts make this the diagnostic worth computing: the spectrum
    lies in [0, 2] for every graph, and the multiplicity of the
    eigenvalue 0 is the number of connected components.  Both are
    returned so a caller can see them rather than assume them.

    Formula: spectrum of Lcal = T^-1/2 (T - W) T^-1/2, lambda_0 = 0
             <= lambda_1 <= ... <= lambda_{n-1} <= 2

    Parameters
    ----------
    W : array-like, shape (n, n)
        Symmetric non-negative weight matrix.
    tol : float
        An eigenvalue below this counts as zero.

    Returns
    -------
    RichResult
        ``values`` (increasing), ``lambda1`` (smallest non-zero),
        ``n_components``, ``max_value``, ``n``.

    References
    ----------
    Chung (1997), Spectral Graph Theory, CBMS 92, Sections 1.2-1.3:
    the eigenvalues of Lcal lie in [0, 2] and the multiplicity of 0
    equals the number of connected components.  Fetched from the
    author's own copy of the chapter.
    """
    W = C.mat(W)
    n = len(W)
    if any(len(r) != n for r in W):
        raise ValueError("W must be square")
    d = [sum(W[i]) for i in range(n)]
    s = [0.0 if d[i] == 0.0 else d[i] ** -0.5 for i in range(n)]
    L = [[(d[i] - W[i][i]) if i == j else -W[i][j] for j in range(n)]
         for i in range(n)]
    Lc = [[s[i] * L[i][j] * s[j] for j in range(n)] for i in range(n)]
    vals, _ = C.eigsym(Lc)
    vals = list(reversed(vals))
    nz = [v for v in vals if v > tol]
    return RichResult(payload={
        "values": vals, "lambda1": nz[0] if nz else float("nan"),
        "n_components": sum(1 for v in vals if v <= tol),
        "max_value": vals[-1], "n": n,
        "method": "Spectrum of the normalised Laplacian"})


sgt_spectrum = lapspec


def cheatsheet():
    return "sgtspc: eigenvalues of Lcal in [0,2]; mult(0) = #components"

File: r-package/morie/R/sgtspc.R
# SPDX-License-Identifier: AGPL-3.0-or-later
#' Spectrum and spectral gap of the normalized Laplacian.
#'
#' Formula: eigenvalues of L = I - D^(-1/2) A D^(-1/2);  lambda_G = least nonzero eigenvalue
#'
#' @param Adj Symmetric adjacency matrix with a zero diagonal; entries may be 0/1 or non-negative edge weights.
#' @param tol Eigenvalues no larger than tol in absolute value count as zero when the spectral gap is picked out.
#'
#' @return List with ``eigenvalues``, ``spectral_gap``, ``n_zero``, ``lambda_max``, ``n``.
#' @references Chung, F. (1997), Spectral Graph Theory, CBMS Regional Conference Series in Mathematics 92, American Mathematical Society, is this shelf's primary book and is NOT in the reference library, so it could not be read.  The conventions below were taken instead from the author's own survey, Chung, F., Four proofs for the Cheeger inequality and graph partition algorithms, Proceedings of ICCM 2007 Vol. II pp. 1-4, Sect. 2 (Preliminaries), which restates them in her notation; that paper was FETCHED and is archived in the reference library with a row in EXTERNAL_SOURCES.md.  Sect. 2: the spectral gap lambda_G is the least NONZERO eigenvalue of the normalized Laplacian.  Its eigenvalues lie in [0, 2] and the multiplicity of the eigenvalue 0 is the number of connected components, so ``n_zero`` counts components.  Eigenvalues are returned in increasing order.
#' @export
Glapspec <- function(Adj, tol = 1e-9) {
  A <- .t1_mat(Adj); n <- nrow(A)
  if (n == 0L || ncol(A) != n) stop("Adj must be a non-empty square matrix")
  if (any(A < 0)) stop("edge weights must be non-negative")
  if (any(A != t(A))) stop("Adj must be symmetric")
  d <- rowSums(A)

  s <- ifelse(d == 0, 0, 1 / sqrt(d))
  L <- diag(n) - (s %o% s) * A
  lam <- sort(.t1_eigsym(L)$values)
  tol <- as.numeric(tol)
  nz <- sum(abs(lam) <= tol)
  gap <- if (any(abs(lam) > tol)) lam[abs(lam) > tol][1] else NA_real_
  .t1_result(eigenvalues = lam, spectral_gap = gap, n_zero = nz,
             lambda_max = lam[length(lam)], n = n,
             method = "Normalized Laplacian spectrum and spectral gap")
}
